### PR TITLE
Added new filters to facilitate ticket scope promotions and new indexes to promotions table

### PR DESCRIPTION
**Pull request by tn3rb**
## Problem this Pull Request solves
Currently there is no UI to facilitate adding ticket scope promotions and doing so will be a lot of work.
This pull request adds a couple of filters to the promotions module so that people with basic coding skills can manually rescope a promotion so that it will be applied to tickets instead of events.

Also added a few indexes to the promotions table for some commonly queried fields which should help speed up those queries

plz also see: https://github.com/eventespresso/event-espresso-core/pull/807

## How has this been tested
monkey based browser testing only

# ticket scope promotions snippet

```php
// PLEASE ADD DATA TO THE FOLLOWING ARRAY
$applicable_promotion_tickets = array();
// this is an array where keys are promotion IDs
// and values are arrays of ticket IDs that are applicable to that promotion
// in the following format: array( promo_ID => array( ticket_ID, ticket_ID ) );
// ex: array(2 => array(30));
// STOP!!! do not edit anything else beyond this
add_filter(
    'FHEE__EED_Promotions__get_applicable_items__applicable_items',
    function ($applicable_items, $promotion) use ($applicable_promotion_tickets) {
        $promotion_IDs = array_keys($applicable_promotion_tickets);
        if (
            ! $promotion instanceof EE_Promotion
            || ! in_array($promotion->ID(), $promotion_IDs, true)
        ) {
            return $applicable_items;
        }
        foreach ($applicable_items as $key => $applicable_item) {
            if ($applicable_item instanceof EE_Line_Item && $applicable_item->OBJ_type() === 'Event') {
                $ticket_line_items = EEH_Line_Item::get_ticket_line_items($applicable_item);
                if (is_array($ticket_line_items)) {
                    foreach ($ticket_line_items as $ticket_line_item) {
                        foreach ($applicable_promotion_tickets as $promotion_ID => $promotion_tickets) {
                            if (
                                $ticket_line_item instanceof EE_Line_Item
                                && $promotion_ID === $promotion->ID()
                                && in_array($ticket_line_item->OBJ_ID(), $promotion_tickets, true)
                            ) {
                                $applicable_items[ $key ] = $ticket_line_item;
                                add_filter(
                                    'FHEE__EED_Promotions__add_promotion_line_item__bypass_increment_promotion_scope_uses',
                                    function (
                                        $bypass_increment_promotion_scope_uses,
                                        $parent_line_item,
                                        $bypass_promotion
                                    ) use ($ticket_line_item, $promotion) {
                                        if ($parent_line_item === $ticket_line_item && $bypass_promotion === $promotion) {
                                            $bypass_increment_promotion_scope_uses = true;
                                        }
                                        return $bypass_increment_promotion_scope_uses;
                                    },
                                    10, 4
                                );
                            }
                        }
                    }
                }
            }
        }
        return $applicable_items;
    },
    10, 3
);
```

### CODE REVIEW

File: core/data_migration_scripts/EE_DMS_Promotions_1_0_0.dms.php [hunk 1]
```diff
@@ -75,7 +75,10 @@ public function schema_changes_before_migration()
                 PRO_deleted tinyint(1) NOT NULL DEFAULT 0,
                 PRO_wp_user bigint(20) unsigned NOT NULL DEFAULT 1,
                 PRIMARY KEY  (PRO_ID),
-                KEY PRC_ID (PRC_ID)",
+                KEY PRC_ID (PRC_ID),
+				KEY PRO_code (PRO_code),
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply core/data_migration_scripts/EE_DMS_Promotions_1_0_0.dms.php]: I'll remove the index changes and create another PR for them

File: EED_Promotions.module.php [hunk 2]
```diff
@@ -738,6 +738,14 @@ public function get_applicable_items(
                 $cart->get_grand_total(),
                 $redeemable_scope_promos
             );
+            $applicable_items = apply_filters(
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply EED_Promotions.module.php]: The location was chosen based on accessibility to variables, primarily the promotion object itself.

**Comment by tn3rb** [comment]
### testing

- ensure that promo codes can still be applied as expected and no errors occur
- plz compare any potential issues against master branch to determine if they are caused by the work in this branch

File: core/data_migration_scripts/EE_DMS_Promotions_1_0_0.dms.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (mnelson4) [reply core/data_migration_scripts/EE_DMS_Promotions_1_0_0.dms.php]: :+1:

**Comment by mnelson4** [comment]
I just added a commit which, from my testing, makes https://github.com/eventespresso/event-espresso-core/pull/807 unnecessary.
You do need to also use the following code snippet too though:

```
add_filter(
    'FHEE__EE_Promotion_Scope__generate_promotion_line_item',
    function(
        $new_line_item_props,
        $parent_line_item,
        $promotion
    ) use ($applicable_promotion_tickets) {
        if($new_line_item_props['OBJ_type'] === 'Promotion'
            && in_array($new_line_item_props['OBJ_ID'],array_keys($applicable_promotion_tickets))){
            $new_line_item_props['LIN_type'] = EEM_Line_Item::type_sub_line_item;
        }
        return $new_line_item_props;
},
    10,
    3
);
```

**Comment by tn3rb** [comment]
yup works great! thnx dude

here's the revised code snippet:

```php
// PLEASE ADD DATA TO THE FOLLOWING ARRAY
$applicable_promotion_tickets = array();
// this is an array where keys are promotion IDs
// and values are arrays of ticket IDs that are applicable to that promotion
// in the following format: array( promo_ID => array( ticket_ID, ticket_ID ) );
// ex: array( 2 => array( 30, 35 ) );
// STOP!!! do not edit anything else beyond this
add_filter(
    'FHEE__EED_Promotions__get_applicable_items__applicable_items',
    function ($applicable_items, $promotion) use ($applicable_promotion_tickets) {
        $promotion_IDs = array_keys($applicable_promotion_tickets);
        if (
            ! $promotion instanceof EE_Promotion
            || ! in_array($promotion->ID(), $promotion_IDs, true)
        ) {
            return $applicable_items;
        }
        foreach ($applicable_items as $key => $applicable_item) {
            if ($applicable_item instanceof EE_Line_Item && $applicable_item->OBJ_type() === 'Event') {
                $ticket_line_items = EEH_Line_Item::get_ticket_line_items($applicable_item);
                if (is_array($ticket_line_items)) {
                    foreach ($ticket_line_items as $ticket_line_item) {
                        foreach ($applicable_promotion_tickets as $promotion_ID => $promotion_tickets) {
                            if (
                                $ticket_line_item instanceof EE_Line_Item
                                && $promotion_ID === $promotion->ID()
                                && in_array($ticket_line_item->OBJ_ID(), $promotion_tickets, true)
                            ) {
                                $applicable_items[ $key ] = $ticket_line_item;
                                add_filter(
                                    'FHEE__EED_Promotions__add_promotion_line_item__bypass_increment_promotion_scope_uses',
                                    function (
                                        $bypass_increment_promotion_scope_uses,
                                        $parent_line_item,
                                        $bypass_promotion
                                    ) use ($ticket_line_item, $promotion) {
                                        if ($parent_line_item === $ticket_line_item
                                            && $bypass_promotion === $promotion
                                        ) {
                                            $bypass_increment_promotion_scope_uses = true;
                                        }
                                        return $bypass_increment_promotion_scope_uses;
                                    },
                                    10, 4
                                );
                                add_filter(
                                    'FHEE__EE_Promotion_Scope__generate_promotion_line_item',
                                    function (
                                        $new_line_item_props,
                                        $parent_line_item,
                                        $promotion
                                    ) use ($promotion_IDs) {
                                        if ($new_line_item_props['OBJ_type'] === 'Promotion'
                                            && in_array(
                                                $new_line_item_props['OBJ_ID'],
                                                $promotion_IDs,
                                                true
                                            )
                                        ) {
                                            $new_line_item_props['LIN_type'] = EEM_Line_Item::type_sub_line_item;
                                        }
                                        return $new_line_item_props;
                                    },
                                    10,
                                    3
                                );
                            }
                        }
                    }
                }
            }
        }
        return $applicable_items;
    },
    10, 3
);
```

**Comment by joshfeck** [comment]
I tried this, but couldn't get it to work. I've switched to this PR's branch. I added the snippet to a plugin and verified that the code is activated. 

I've set promo ID 36 to be applicable to only ticket IDs 770 & 772
`$applicable_promotion_tickets = array( 36 => array(770,772));`

Then I go to register for ticket ID 776 **only** and at checkout I input the promo code for promo ID 36. I'd expect the promo code to be rejected, but it's accepted. 

Is there something I could be missing to make this work?

**Comment by mnelson4** [comment]
Thanks Josh.
If you setup a promotion (with or without a discount code, I don't think it matters) to apply to ticket A, but not ticket B, then purchase BOTH simulatneously, I'm seeing this work as expected: the promotion only applies to ticket A.

HOWEVER, if you only purchase ticket B, to which the discount should NOT apply, according to `$applicable_promotion_tickets `, it unfortunately does.

I think @tn3rb might have a better idea of why that is. My contributions for this were mostly regarding line item structure, whereas this has more to do with deciding when to apply the promotion.

Also, with this code snippet, the "ticket-scoped" promotions need to also apply to the event you're registering for. I'm not sure if that's a feature or a bug.

**Comment by joshfeck** [comment]
> If you setup a promotion (with or without a discount code, I don't think it matters) to apply to ticket A, but not ticket B, then purchase BOTH simultaneously, I'm seeing this work as expected: the promotion only applies to ticket A.

Was that with a percentage discount or a fixed/dollar discount? If the latter, the end result would be equivalent without the filter.

**Comment by mnelson4** [comment]
Thanks Josh, I tried it with a percent.

[Here's when I registered for BOTH tickets](https://drive.google.com/a/eventespresso.com/file/d/1pkUyI14vyMxbZ39Y5VbqG4bS9dZzCfrx/view?usp=drivesdk) (same event). "Discountable" got the discount :+1: (because its in `$applicable_promotion_tickets`), and "NoDiscountEver" didn't get the discount :+1: (because it's not in `$applicable_promotion_tickets`). The code snippet worked there.

[BUT, here's where I only registered for "NoDiscountEver"](https://drive.google.com/a/eventespresso.com/file/d/1f2Hp8z4yqTtPn0UNcL0Yq-cqXi30456j/view?usp=drivesdk) (not in `$applicable_promotion_tickets`). It erroneously got discounted.

**Comment by tn3rb** [comment]
oops... ya looks like O forgot to add logic to exclude other tickets.. will fix

**Comment by tn3rb** [comment]
@joshfeck 
here's the updated snippet, want to give it another try plz?

```php

// PLEASE ADD DATA TO THE FOLLOWING ARRAY
$applicable_promotion_tickets = array();
// this is an array where keys are promotion IDs
// and values are arrays of ticket IDs that are applicable to that promotion
// in the following format: array( promo_ID => array( ticket_ID, ticket_ID ) );
// ex: array( 2 => array( 30, 35 ) );
// STOP!!! do not edit anything else beyond this
add_filter(
    'FHEE__EED_Promotions__get_applicable_items__applicable_items',
    function ($applicable_items, $promotion) use ($applicable_promotion_tickets) {
        $promotion_IDs = array_keys($applicable_promotion_tickets);
        if (
            ! $promotion instanceof EE_Promotion
            || ! in_array($promotion->ID(), $promotion_IDs, true)
        ) {
            return $applicable_items;
        }
        foreach ($applicable_items as $key => $applicable_item) {
            if ($applicable_item instanceof EE_Line_Item && $applicable_item->OBJ_type() === 'Event') {
                $ticket_line_items = EEH_Line_Item::get_ticket_line_items($applicable_item);
                $valid_items = [];
                $invalid_items = [];
                if (is_array($ticket_line_items)) {
                    foreach ($ticket_line_items as $ticket_line_item) {
                        if (! $ticket_line_item instanceof EE_Line_Item) {
                            continue;
                        }
                        foreach ($applicable_promotion_tickets as $promotion_ID => $promotion_tickets) {
                            if ($promotion_ID !== $promotion->ID()) {
                                continue;
                            }
                            if (in_array($ticket_line_item->OBJ_ID(), $promotion_tickets, true)) {
                                // overwrite applicable event line item with ticket line item
                                $applicable_items[ $key ] = $ticket_line_item;
                                // and track the key so it's not removed on subsequent iterations
                                $valid_items[] = $key;
                                add_filter(
                                    'FHEE__EED_Promotions__add_promotion_line_item__bypass_increment_promotion_scope_uses',
                                    function (
                                        $bypass_increment_promotion_scope_uses,
                                        $parent_line_item,
                                        $bypass_promotion
                                    ) use ($ticket_line_item, $promotion) {
                                        if ($parent_line_item === $ticket_line_item
                                            && $bypass_promotion === $promotion
                                        ) {
                                            $bypass_increment_promotion_scope_uses = true;
                                        }
                                        return $bypass_increment_promotion_scope_uses;
                                    },
                                    10, 4
                                );
                                add_filter(
                                    'FHEE__EE_Promotion_Scope__generate_promotion_line_item',
                                    function ($new_line_item_props) use ($promotion_IDs) {
                                        if ($new_line_item_props['OBJ_type'] === 'Promotion'
                                            && in_array(
                                                $new_line_item_props['OBJ_ID'],
                                                $promotion_IDs,
                                                true
                                            )
                                        ) {
                                            $new_line_item_props['LIN_type'] = EEM_Line_Item::type_sub_line_item;
                                        }
                                        return $new_line_item_props;
                                    }
                                );
                            } else {
                                // this ticket is not valid, but don't remove the applicable item just yet
                                $invalid_items[] = $key;
                            }
                        }
                    }
                }
                // remove valid items from list of invalid ones
                $invalid_items = array_diff($invalid_items, $valid_items);
                // then remove invalid items from list of applicable items
                foreach ($invalid_items as $invalid_item) {
                    unset($applicable_items[ $invalid_item ]);
                }
            }
        }
        return $applicable_items;
    },
    10, 3
);
```

**Pull request merged**